pkg/tso/keyspace_group_manager.go

```diff
@@ -495,39 +505,86 @@ func (kgm *KeyspaceGroupManager) deleteKeyspaceGroup(id uint32) {
-	if err := kgm.checkKeySpaceGroupID(id); err != nil {
+func (kgm *KeyspaceGroupManager) GetAllocatorManager(keyspaceGroupID uint32) (*AllocatorManager, error) {
+	if err := kgm.checkKeySpaceGroupID(keyspaceGroupID); err != nil {
 		return nil, err
 	}
-	if am := kgm.ams[id].Load(); am != nil {
+	if am := kgm.ams[keyspaceGroupID].Load(); am != nil {
+		return am, nil
+	}
+	return nil, kgm.genNotServedErr(errs.ErrGetAllocatorManager, keyspaceGroupID)
+}
+
+// GetAMWithMembershipCheck returns the AllocatorManager of the given keyspace group and check if the keyspace
+// is served by this keyspace group.
+func (kgm *KeyspaceGroupManager) GetAMWithMembershipCheck(
+	keyspaceID, keyspaceGroupID uint32,
+) (*AllocatorManager, error) {
+	if am := kgm.ams[keyspaceGroupID].Load(); am != nil {
+		ksg := kgm.ksgs[keyspaceGroupID].Load()
+		if ksg == nil {
+			return nil, kgm.genNotServedErr(errs.ErrGetAllocatorManager, keyspaceGroupID)
+		}
+		if _, ok := ksg.KeyspaceLookupTable[keyspaceID]; !ok {
+			return nil, kgm.genNotServedErr(errs.ErrGetAllocatorManager, keyspaceGroupID)
+		}
 		return am, nil
 	}
-	return nil, errs.ErrGetAllocatorManager.FastGenByArgs(
-		fmt.Sprintf("requested keyspace group with id %d %s by this host/pod", id, errs.NotServedErr))
+	return nil, kgm.genNotServedErr(errs.ErrGetAllocatorManager, keyspaceGroupID)
 }
 
 // GetElectionMember returns the election member of the given keyspace group
-func (kgm *KeyspaceGroupManager) GetElectionMember(id uint32) (ElectionMember, error) {
-	am, err := kgm.GetAllocatorManager(id)
+// TODO: support multiple keyspace groups for GetElectionMember
+func (kgm *KeyspaceGroupManager) GetElectionMember(
+	keyspaceID, keyspaceGroupID uint32,
+) (ElectionMember, error) {
+	if err := kgm.checkKeySpaceGroupID(keyspaceGroupID); err != nil {
+		return nil, err
+	}
+	am, err := kgm.GetAMWithMembershipCheck(keyspaceID, keyspaceGroupID)
 	if err != nil {
 		return nil, err
 	}
 	return am.getMember(), nil
 }
 
 // HandleTSORequest forwards TSO allocation requests to correct TSO Allocators of the given keyspace group.
-func (kgm *KeyspaceGroupManager) HandleTSORequest(id uint32, dcLocation string, count uint32) (pdpb.Timestamp, error) {
-	am, err := kgm.GetAllocatorManager(id)
+func (kgm *KeyspaceGroupManager) HandleTSORequest(
+	keyspaceID, keyspaceGroupID uint32,
+	dcLocation string, count uint32,
+) (ts pdpb.Timestamp, currentKeyspaceGroupID uint32, err error) {
+	if err := kgm.checkKeySpaceGroupID(keyspaceGroupID); err != nil {
+		return pdpb.Timestamp{}, keyspaceGroupID, err
+	}
+	am, err := kgm.GetAMWithMembershipCheck(keyspaceID, keyspaceGroupID)
 	if err != nil {
-		return pdpb.Timestamp{}, err
+		// The keyspace doesn't belong to this keyspace group, we should check if it belongs to any other
+		// keyspace groups, and return the correct keyspace group ID to the client.
+		for i := 0; i < int(mcsutils.MaxKeyspaceGroupCountInUse); i++ {
+			if ksg := kgm.ksgs[i].Load(); ksg == nil {
```

Comment:
Will it cost too much?

Reply:
This is the rare case, but I reimplement with new algorithm to reduce time complexity from O(mlogn) to O(logn), where m is the number of keyspace groups and n the number of keyspaces.